core/cache/model.go

```diff
@@ -0,0 +1,141 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package cache
+
+import (
+	"sync"
+
+	"gopkg.in/tomb.v2"
+
+	"github.com/juju/juju/core/watcher"
+	"github.com/juju/pubsub"
+)
+
+const modelConfigChange = "model-config-change"
+
+// Model is a cached model in the controller. The model is kept up to
+// date with changes flowing into the cached controller.
+type Model struct {
+	hub *pubsub.SimpleHub
+	mu  sync.Mutex
+
+	details    ModelChange
+	configHash string
+}
+
+// modelTopic prefixes the topic with the model UUID.
+func (m *Model) modelTopic(topic string) string {
+	return m.details.ModelUUID + ":" + topic
+}
+
+func (m *Model) setDetails(details ModelChange) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	m.details = details
+
+	configHash, err := hash(details.Config)
+	if err != nil {
+		logger.Errorf("invariant error - model config should be yaml serializable and hashable, %v", err)
+		return
+	}
+	if configHash != m.configHash {
+		m.configHash = configHash
+		m.hub.Publish(m.modelTopic(modelConfigChange), details.Config)
+	}
+}
+
+// Config returns the current model config.
+func (m *Model) Config() map[string]interface{} {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	return m.details.Config
+}
+
+// WatchConfig creates a watcher for the model config.
+// If keys are specified, the watcher is only signals a change when
+// those keys change values.
+func (m *Model) WatchConfig(keys ...string) watcher.NotifyWatcher {
+	// We use a single entry buffered channel for the changes.
+	// This allows the config changed handler to send a value when there
+	// is a change, but if that value hasn't been consumed before the
+	// next change, the second change is discarded.
+	watcher := &modelConfigWatcher{
+		keys:    keys,
```

Comment:
Should we allow a zero-length keys, which would never get any changes?
Should it be taken to mean all keys, which might be abused?
Should it be an error?

Reply:
A zero length keys means watch for any changes. Added more doc.